statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,336 @@
+	l.RLock()
+	defer l.RUnlock()
+	r := make(map[int64]*statistics.Table, len(l.elements))
+	for k, v := range l.elements {
+		r[k] = v.Value.(*cacheItem).value
+	}
+	return r
+}
+
+// Len returns the current length
+func (l *internalLRUCache) Len() int {
+	l.RLock()
+	defer l.RUnlock()
+	return len(l.elements)
+}
+
+// FreshMemUsage re-calculate the memory message
+func (l *internalLRUCache) FreshMemUsage() {
+	l.Lock()
+	for _, v := range l.elements {
+		item := v.Value.(*cacheItem)
+		oldMemUsage := item.tblMemUsage
+		newMemUsage := item.value.MemoryUsage()
+		item.tblMemUsage = newMemUsage
+		l.calculateCost(newMemUsage, oldMemUsage)
+		l.maintainList(v, nil, newMemUsage.TotalColTrackingMemUsage(), oldMemUsage.TotalColTrackingMemUsage())
+	}
+	l.evictIfNeeded()
+	totalCost := l.totalCost
+	trackingCost := l.trackingCost
+	l.Unlock()
+	metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+	metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+}
+
+// FreshTableCost re-calculate the memory message for the certain key
+func (l *internalLRUCache) FreshTableCost(key int64) {
+	l.Lock()
+	calculated := l.calculateTableCost(key)
+	totalCost := l.totalCost
+	trackingCost := l.trackingCost
+	l.Unlock()
+	if calculated {
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+	}
+}
+
+func (l *internalLRUCache) calculateTableCost(key int64) bool {
+	element, exists := l.elements[key]
+	if !exists {
+		return false
+	}
+	item := element.Value.(*cacheItem)
+	l.put(item.key, item.value, item.value.MemoryUsage(), true)
+	return true
+}
+
+// Copy returns a replication of LRU
+func (l *internalLRUCache) Copy() statsCacheInner {
+	var newCache *internalLRUCache
+	l.RLock()
+	newCache = newInternalLRUCache(l.capacity)
+	node := l.cache.Back()
+	for node != nil {
+		key := node.Value.(*cacheItem).key
+		value := node.Value.(*cacheItem).value
+		tblMemUsage := node.Value.(*cacheItem).tblMemUsage
+		newCache.put(key, value, tblMemUsage, false)
+		node = node.Prev()
+	}
+	l.RUnlock()
+	metrics.StatsCacheLRUCounter.WithLabelValues(typCopy).Inc()
+	return newCache
+}
+
+// internalLRUCache will evict a table's column' structure in order to keep tracking cost under capacity
+// If the elements has no structure can be evicted, it will be removed from list.
+func (l *internalLRUCache) evictIfNeeded() {
+	curr := l.cache.Back()
+	evicted := false
+	for !l.underCapacity() {
+		evicted = true
+		item := curr.Value.(*cacheItem)
+		tbl := item.value
+		oldMemUsage := item.tblMemUsage
+		prev := curr.Prev()
+		for _, col := range tbl.Columns {
+			if col.IsEvicted() {
+				continue
+			}
+			col.DropEvicted()
+			newMemUsage := tbl.MemoryUsage()
+			item.tblMemUsage = newMemUsage
+			l.calculateCost(newMemUsage, oldMemUsage)
+			if l.underCapacity() {
+				break
```

Comment:
Since we are maintaining LRU at Table level, why not evict CMSketch for all columns of this table directly to make it simple.